src/doughnut/mod.rs

```diff
@@ -0,0 +1,66 @@
+// Copyright 2019 Centrality Investments Limited
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//!
+//! A versioned doughnut wrapper.
+//!
+
+#![allow(clippy::type_repetition_in_bounds)]
+
+use crate::v0::parity::DoughnutV0;
+use codec::{Decode, Encode, Error};
+use core::convert::TryFrom;
+
+/// A versioned doughnut wrapper. It proxies to the real,inner doughnut type
+#[derive(Encode, Decode, Clone, Debug, Eq, PartialEq)]
+pub enum Doughnut {
+    V0(DoughnutV0),
+}
+
+#[allow(unreachable_patterns)]
+impl TryFrom<Doughnut> for DoughnutV0 {
+    type Error = Error;
+    fn try_from(v: Doughnut) -> Result<Self, Self::Error> {
+        match v {
```

Comment:
Can change this to an `if let` and get rid of the unreachable patterns attr
```rust
if let Doughnut::V0(inner) = v {
    return Ok(inner)
}
return Err(..)
```

Reply:
Currently we only have one item in versioned_doughnut, need to add `#[allow(irrefutable_let_patterns)]
` if use `if let`